### ISSUE TITLE
Optional arguments in DU constructors

### ISSUE THREAD

**Issue by charlesroddie** [fsharp/fslang-suggestions#854]
Often in DU cases there are settings that sometimes need specifying and are represented as option types:

```fsharp
type TextElement =
    | String of text:string * fontSize: float option * color: Color option
    | PageBreak
let s1 = String("Normal text.", None, None)
let s2 = String("Blue text.", None, Some Color.Blue)
```

These should be writable using optional argument constructor syntax:
```fsharp
type TextElement =
    | String of text:string * ?fontSize: float * ?color: Color
    | PageBreak
let s1 = String("Normal text.")
let s2 = String("Blue text.", color = Color.Blue)
```

Related suggestions: https://github.com/fsharp/fslang-suggestions/issues/852 and related, which would allow making DU constructors private and implementing separate constructors as static methods with optional arguments.

**Comment by Tarmil** [comment]
I like this idea. How would pattern matching work? Something like this would probably be the most consistent with optional method arguments:

```fsharp
match x with
// Matching without name => only matches if it's Some, binds the value
// Not mentioned => always matches (ie color in all the examples)
| String(_, size) -> printfn "fontSize is %f" size

// Matching with name => only matches if it's Some, binds the value
| String(_, fontSize = size) -> printfn "fontSize is %f" size
| String _ -> printfn "No fontSize"

// Matching with ?name => binds the option
| String(_, ?fontSize = None) -> printfn "fontSize is unspecified"
```

**Comment by charlesroddie** [comment]
> How would pattern matching work?

I would do this in a way that only affects constructors and doesn't affect pattern matching. Similarly to the other suggestions relating to access modifiers on constructors, we have to separate the constructors and the matchable data.

**Comment by Tarmil** [comment]
I think it would feel inconsistent that if you use `String(text = t, fontSize = f, color = c)` in the constructor, then `f` and `c` have type `float` and `Color`, but with the same syntax in pattern matching, they're `float option` and `Color option`.

**Comment by TheJayMann** [comment]
Having the `f` and `c` be `float` and `Color`, while having `fontSize` and `color` be `float option` and `Color option` in pattern matching is just as consistent as having a method defined which takes optional arguments, where the method takes in values which are not options, but the method body sees them as options.

Also, you can recursively match against `Some` and `None` when using pattern matching.

```fsharp
match x with
| String(_, Some size, _) -> printfn "fontSize is %f" size

| String(_, Some fontSize, _) -> printfn "fontSize is %f" size
| String(_, _, _) -> printfn "No fontSize"

| String(_, None as fontSize, _) -> printfn "fontSize is unspecified"
```

**Comment by charlesroddie** [comment]
Than usual option argument syntax has this type of inconsistency. For example:
```fsharp
type X(?y:int) =
    member _.Y = y
X(y=0).Y // Some 0: y is specified by `y=0` but is an int option when accessed.
```
As @TheJayMann points out this is not present in DU matching. You can match against any name you like.

**Comment by realparadyne** [comment]
Doesn't a DU case have zero or one items of data associated with it?  And that one item can be a tuple.  So if this change were made, wouldn't it really be tuples that are getting changed rather than specifically just DU cases?

Either way, the compiler would no longer warn me in cases where I've forgotten to consider a piece of data in the tuple as it would assume I just wanted to save typing.

Worse, when someone else adds a new element into the tuple definition, they won't get any warnings elsewhere in the code where it should at least be considered if None is the right value to now be using to construct the tuple or DU case.

I think this change would lead F# away from correctness and safety, which are some of its core strengths.

**Comment by Tarmil** [comment]
> Doesn't a DU case have zero or one items of data associated with it? And that one item can be a tuple. So if this change were made, wouldn't it really be tuples that are getting changed rather than specifically just DU cases?

No, tuple items look like a tuple syntactically but they're proper separate fields of the DU case.

> Either way, the compiler would no longer warn me in cases where I've forgotten to consider a piece of data in the tuple as it would assume I just wanted to save typing.
> 
> Worse, when someone else adds a new element into the tuple definition, they won't get any warnings elsewhere in the code where it should at least be considered if None is the right value to now be using to construct the tuple or DU case.

Only if the new field was explicitly declared as optional with a `?`. And if it was, then it was *intentionally* non-breaking. But adding a field with the usual syntax would continue to have the current behavior. I agree that it would be terrible if it didn't!
